### ISSUE TITLE
Accessibility support for custom actions

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#18432]
cc @DaveShuckerow 

On Android, a re-orderable list will also present reorder options in the local context menu so that they can be accessed via alternative inputs.  For an example, see the language selection in the Settings app. iOS presents similar options

We could expose an API to add local context actions which would need to callback into dart code. Alternatively we could only support a predefined list of local context actions and add them to the semantics widget.

To be as backwards compatible as possible, the implementation on the Android side would use the support libraries, but this seems to be blocked by flutter/flutter#11099

Relevant Android APIs:
https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction

Relevant iOS APIs:
https://developer.apple.com/documentation/uikit/accessibility/uiaccessibilityaction#//apple_ref/occ/instp/NSObject/accessibilityCustomActions

**Comment by jonahwilliams** [comment]
@DaveShuckerow if the reorder list actions are a blocker feel free to update this issue

**Comment by jonahwilliams** [comment]
This is also needed by a leave-behind style list

**Comment by jonahwilliams** [comment]
To support these we would need a way for Flutter to provide the engine with a display name and identifier for each custom action on the semantics node that supported it.  The engine would take this information and construct the relevant classes.  When a semantics action called back into flutter, we would need a bit/bundle info which indicated the identifier of the custom action.

Issues:

### Encoding custom actions in the semantics tree.
Currently we use a bitmap to send which actions are enabled on a semantics node.  If we support arbitrary custom actions we would need to additionally send a String display name and an identifier (which may be the same as the name) on any semantics nodes which support them.


### Platform support for custom actions
Creating custom actions requires android sdk version >= 21.  We can get some support on earlier versions using compat libraries.

### Invoking custom actions
We could reserve a bit on the current actions which indicates "custom action", so that the framework would know to look for the semantic callback in a different place (on the framework side).

In the Android accessibility bridge we can use the default switch case to handle all custom actions, they don't need to all be known statically.  But we need a way to map the id used for this with the id of the custom action.


## Proposed API
create a custom semantics action and provide to an additional property of the semantics widget, along with the callback to be invoked.

Dart
```dart
/// Creates a new custom semantics action with a unique ID assigned by the framework.
class CustomSemanticsAction implements SemanticsAction {
  CustomSemanticsAction(String label);
}

...

const CustomSemanticsAction moveUpAction = const CustomSemanticsAction('move up');

Semantics(customActions: <CustomSemanticsAction, Callback>{
  moveUpAction: () { ... },
  ...,
);
```

This would get transmitted in the semantics as a list of strings, 

```
SemanticNode(
   ...,
   additionalActions: ['move up']
)
```

When the AccessibilityBridge receives a node with a custom action.  It creates a (cached) instance of the respective class and uses the platform APIs to register this with the semantics node.  When this action is invoked it passes an action with a specific flag CUSTOM_ACTION, and an argument of this custom actions name.  The Framework semantics logic looks in the separate map for the specified callback and invokes it if it exists.

**Comment by jonahwilliams** [comment]
cc @Hixie and @goderbauer for semantics API

**Comment by goderbauer** [comment]
Is there an equivalent for these custom actions on iOS? What information do we need to expose them there?

Would it make sense to consider this proposal together with https://github.com/flutter/flutter/issues/14402? I.e. you can "subclass" one of the build-in actions to provide a customized hint for those actions?

**Comment by jonahwilliams** [comment]
https://developer.apple.com/documentation/uikit/uiaccessibilitycustomaction - found thanks to this helpful NPR story https://npr.codes/custom-accessibility-actions-for-ios-3e35409c9dcc

**Comment by jonahwilliams** [comment]
I'm not sure if existing actions are customizable on iOS.  Even if the implementation is similar on Android I think they should be different APIs
